### PR TITLE
Use the MSBuild Pkg<package-id> property for package location (#36)

### PR DESCRIPTION
**Pull request by jeffrosenberg**
For packages that meet certain criteria (such as having a tools folder), MSBuild will automatically include a property that resolves their location. This is more accurate because it handles floating versions$

### CODE REVIEW

**Comment by jeffrosenberg** [comment]
Looks like this fails on dotnet 2.1, with the error `error : '1.*-*' is not a valid version string.` That makes sense, as it's a relatively recent change. I'll need to give some thought to the best way to clean up the tests. I think it would be valid to test the floating versions differently on 2.1 and 3.1, but I don't want to complicate the tests too much.

**Comment by jmezach** [comment]
Yeah, that makes sense. I guess we can make the `PackageReference` conditional based on some property which we pass in on the command line. Then we can have the 3.1 jobs pass that property, while we don't pass it for the 2.1 jobs.

**Comment by jeffrosenberg** [comment]
@jmezach makes sense to me. I have a thought on how to do that, but I can't test it locally so apologies for what's likely to be a few rounds of pushing changes and running tests.

**Comment by jmezach** [comment]
You can probably test it locally by having the latest 2.1 SDK installed and putting a global.json somewhere to force that version to be used. Not that I mind testing it through CI though, but I think it's always easier to test locally first ;).

**Comment by jeffrosenberg** [comment]
The issue I had is testing some of the conditionals in the Github actions file. What I've done here is to expand the use of `cschleiden/replace-tokens` in a few ways:

* Add a new `#{DependencyVersion}#` placeholder
* For dotnet 3.1, replace `#{DependencyVersion}#` with a double wildcard: `1.*-*`
* For dotnet 2.1, take a two-step approach:
  1. First replace `#{DependencyVersion}#` with another token expression: `#{NBGV_Version}#-*`
  2. Then replace the `#{NBGV_Version}#` token in the next step to resolve the major/minor/patch version

File: src/MSBuild.Sdk.SqlProj/Sdk/Sdk.targets
Comment (jmezach) [inline src/MSBuild.Sdk.SqlProj/Sdk/Sdk.targets:115]: I'm guessing you got this comment from my original commit since this is referring to an internal package here. We should probably replace `DeVries.WMS.Database` here with something like `Foo.Bar.Database` or something.

File: test/TestProjectWithSDKRef/TestProjectWithSDKRef.csproj
Comment (jmezach) [inline test/TestProjectWithSDKRef/TestProjectWithSDKRef.csproj:12]: What if we were to define two `PackageReference`'s here, one with `Condition="'$(UseFloatingVersions)'=='True'"` and one with `!=True` and then have two different tokens (one using a floating version and one without). Then we would only have to add `/p:UseFloatingVersions=True` to the `dotnet build` command depending on which SDK we're currently using?

**Review by jmezach — CHANGES_REQUESTED** [review]
Looks good to me overall. I want to dig into the binary log to make sure it is working as expected before we merge this though.

**Comment by jmezach** [comment]
Just out of curiosity, have you checked if adding `GeneratePathProperty=False` on the `PackageReference` forces the property not to be generated (even though the package has a `tools` folder)? Then we could use that to test the fallback scenario as well.

File: src/MSBuild.Sdk.SqlProj/Sdk/Sdk.targets
Comment (jeffrosenberg) [reply src/MSBuild.Sdk.SqlProj/Sdk/Sdk.targets:115]: Okay, sure. I wasn't sure if it was an example or a real-world use case, although I suppose it is quite specific 😄

**Comment by jeffrosenberg** [comment]
> Just out of curiosity, have you checked if adding `GeneratePathProperty=False` on the `PackageReference` forces the property not to be generated (even though the package has a `tools` folder)?

Just tried this, and it doesn't work, but I see your point about making sure to test the fallback.

File: test/TestProjectWithSDKRef/TestProjectWithSDKRef.csproj
Comment (jeffrosenberg) [reply test/TestProjectWithSDKRef/TestProjectWithSDKRef.csproj:12]: Are you thinking that `UseFloatingVersions` property would be just for the test project? If so, I think that could make sense. I'd probably tweak the name though, as both versions technically use floating versions, it's just a question of single vs. double wildcards.

**Comment by jeffrosenberg** [comment]
Haha, I figured out that I can apparently test the fallback scenario by redefining the `$(Pkg<package-id>)` property:

`dotnet build ./test/TestProjectWithPackageReference/TestProjectWithPackageReference.csproj -c Release /p:PkgTestProject=""`

I'll work on adding this to the ci tests.

File: test/TestProjectWithSDKRef/TestProjectWithSDKRef.csproj
Comment (jmezach) [reply test/TestProjectWithSDKRef/TestProjectWithSDKRef.csproj:12]: Yeah, sure, I think that should be something we'd use for testing only. I don't expect users to do it this way.

**Comment by jeffrosenberg** [comment]
@jmezach okay, I've made the following changes to the tests, and all are passing:

* Use a conditional property to manage different test scenarios
  * I decided to call it `TestWildcard`, which can be set to "Single", "Double", or left empty
* Run tests on both floating and exact versions
  * One subset of tests for exact versions should be using the fallback method to derive the path
  * I wanted to add a negative test for this as well, i.e. a test expecting failure, but that doesn't really work with the current setup
* Add a flag to treat SQL71502 warnings as errors
  * These are the warnings that pop up when the dependencies aren't resolved correctly. 
  * On my last PR, we were scanning for this scenario manually. This flag should fail the build instead.

File: test/TestProjectWithSDKRef/TestProjectWithSDKRef.csproj
Comment (jmezach) [inline test/TestProjectWithSDKRef/TestProjectWithSDKRef.csproj:12]: Now that I see this I'm thinking, can't we specify the Version we want to use entirely from the command line and test our various scenario's that way? So we just put `Version="$(DependencyVersion)"` in here (with a single `PackageReference` element again) and then control the `DependencyVersion` property from the command line. That would make this a lot cleaner imho, but I am sorry for changing my mind about this every time I see it ;).

**Comment by jeffrosenberg** [comment]
I pushed one small change, just cleaning up the names of the test steps now that they don't have to be separated out by OS. I'm good to merge this one if you guys are. Thanks as always for working with me on this! I'm getting really close to finalizing the integration of this SDK into the database builds at my company! 🙌

**Comment by jmezach** [comment]
@jeffrosenberg Thanks for your contribution. Really appreciate it. And great to hear that you're working on integrating this at your company.

**Pull request merged**
